lib/pa11y.js

```diff
@@ -0,0 +1,154 @@
+//
+// This is pretty naive — it would be better to, say, scan pages
+// that use a changed partial or layout.
+//
+// This scan happens in addition to, not in place of, the
+// pa11yPerPage scan.
+//
+
+function changedFileSet() {
+  const differ = process.env.CI ? new CommitDiffer : new Differ
+  return new Set(differ.changedFiles());
+}
+
+function sitewideFileSet() {
+  const sitewideFolders = ['assets', '_includes', '_sass']
+  const sitewideFiles = [];
+  sitewideFolders.forEach((folder) => {
+    sitewideFiles.push(glob.sync(`${folder}/**/*`))
+  });
+  return new Set(sitewideFiles.flat())
+}
+
+function intersection(setA, setB) {
+  const result = new Set();
+  setA.forEach((value) => {
+    if (setB.has(value)) {
+      result.add(value)
+    }
+  });
+  return result
+}
+
+function doesIntersect(setA, setB) {
+  return intersection(setA, setB).size > 0
+}
+
+function didSitewideFilesChange() {
+  const result = doesIntersect(changedFileSet(), sitewideFileSet())
+  if (result) {
+    console.log('[pa11y] Sitewide files changed, doing sitewide sample') /* eslint-disable-line no-console */
+  } else {
+    console.log('[pa11y] Sitewide files did not change, not doing sitewide sample') /* eslint-disable-line no-console */
+  }
+  return result
+}
+
+function sample(array, limit = 1) {
+  const indices = [];
+  while(indices.length < limit) {
+    const r = Math.floor(Math.random() * (array.length - 1)) + 1;
+    if (indices.indexOf(r) === -1) indices.push(r);
+  }
+  const sampleSet = []
+  indices.forEach((ind) => { sampleSet.push(array[ind]) });
+  if (sampleSet.length === 1) { return sampleSet[0] }
+  return sampleSet
+}
+
+function sampleFiles () {
+  const collections = {
+    authors: '_site/author/*/index.html',
+    blog: '_site/blog/page/[0-9]*/index.html',
+    posts: '_site/20[0-9][0-9]/**/index.html',
+    services: '_site/project/*/index.html',
+    tags: '_site/tags/*/index.html',
+    root: '_site/*/index.html'
+  }
+
+  const siteFiles = []
+  Object.entries(collections).forEach(([name, pattern]) => {
+    if (name === 'root') {
+      siteFiles.push(glob.sync(pattern))
+    } else {
+      siteFiles.push(sample(glob.sync(pattern), 3))
+    }
+  });
+  return siteFiles.flat()
+}
+
+// Looks through all the files coming from eleventy:after. If any of
+// the documents should be scanned, list the files to be scanned in
+// the pa11y targets file.
+//
+// We list the output paths because that's where the generated HTML
+// lives, the page that will be seen by the public.
+//
+function pa11yPerPage({ dir, results }) {
+  const differ = process.env.CI ? new CommitDiffer : new Differ
+  const changedFiles = differ.changedFiles();
+  const filesToWrite = [];
+
+  results.forEach((result) => {
+    const doc = new Document(result, dir, changedFiles);
+    if (doc.shouldScan()) { filesToWrite.push(result.outputPath) }
+  });
+
+  return filesToWrite
+}
+
+async function pa11yScan ({ dir, results }) {
```

Comment:
Non blocking question/suggestion: `results` sounds like it's an array of pages/files/documents sent from `eleventy.after`, if that's correct then maybe renaming `results` to have a little more context could help readability? If so I think the same renaming should occur where `results` is passed to the `pa11yScan` function.